src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp

```diff
@@ -33,63 +32,72 @@ WriteBufferFromAzureBlobStorage::~WriteBufferFromAzureBlobStorage()
     finalize();
 }
 
-void WriteBufferFromAzureBlobStorage::finalizeImpl()
+void WriteBufferFromAzureBlobStorage::execWithRetry(std::function<void()> func, size_t num_tries)
 {
-    if (attributes.has_value())
+    auto can_retry_exception = [&]<typename Exception>(const Exception & e, size_t i) -> bool
     {
-        auto blob_client = blob_container_client->GetBlobClient(blob_path);
-        Azure::Storage::Metadata metadata;
-        for (const auto & [key, value] : *attributes)
-            metadata[key] = value;
-        blob_client.SetMetadata(metadata);
-    }
+        if (i == num_tries - 1)
+            return false;
 
-    const size_t max_tries = 3;
-    for (size_t i = 0; i < max_tries; ++i)
+        LOG_DEBUG(log, "Write at attempt {} for blob `{}` failed: {}", i + 1, blob_path, e.Message);
+        return true;
+    };
+
+    for (size_t i = 0; i < num_tries; ++i)
     {
         try
         {
-            next();
+            func();
             break;
         }
+        catch (const Azure::Core::Http::TransportException & e)
+        {
+            if (!can_retry_exception(e, i))
+                throw;
```

Comment:
maybe have `handle_exception(e, i)` which will throw if retries are used and just log if there are more retries?